File: 0897-increasing-order-search-tree/0897-increasing-order-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    private TreeNode head = null, parent = null;
    public TreeNode increasingBST(TreeNode root) {
        solve(root);
        return head;
    }
    
    private void solve(TreeNode root){
        if(root == null) return;
        solve(root.left);
        if(head == null) head = root;
        if(parent == null) parent = root;
        else {
            parent.right = root;
            parent = root;
            root.left = null;
        }
        solve(root.right); 
    }
}

File: 0897-increasing-order-search-tree/0897-increasing-order-search-tree.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun increasingBST(root: TreeNode?): TreeNode? {
        if (root == null) return null
        
        val stack = Stack<TreeNode>()
        pushAllLeft(root, stack)
        var newRoot: TreeNode? = null
        var pointer: TreeNode?  = null
        while (stack.isNotEmpty()) {
            val node = stack.pop()

            if (newRoot == null) {
                newRoot = TreeNode(node.`val`)
                pointer = newRoot
            } else {
                pointer!!.right = TreeNode(node.`val`)
                pointer = pointer!!.right
            }

            pushAllLeft(node.right, stack)
        }
        return newRoot
    }

    private fun pushAllLeft(root: TreeNode?, stack: Stack<TreeNode>) {
        var p = root
        while (p != null) {
            stack.push(p)
            p = p.left
        }
    }
}